File: android/src/main/kotlin/com/mapbox/maps/mapbox_maps/ViewAnnotationController.kt
package com.mapbox.maps.mapbox_maps

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.widget.TextView
import com.mapbox.geojson.Point
import com.mapbox.maps.MapView
import com.mapbox.maps.ViewAnnotationAnchor
import com.mapbox.maps.viewannotation.ViewAnnotationManager
import com.mapbox.maps.viewannotation.annotationAnchor
import com.mapbox.maps.viewannotation.geometry
import com.mapbox.maps.viewannotation.viewAnnotationOptions
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodCall

class ViewAnnotationController(private val mapView: MapView, private val context: Context) {

  private val viewAnnotationManager: ViewAnnotationManager = mapView.viewAnnotationManager
  private val annotationsMap = mutableMapOf<String, View>() // Map to track annotations by ID

  fun addViewAnnotation(call: MethodCall, result: MethodChannel.Result) {
    val viewAnnotationId = call.argument<String>("viewAnnotationId")
    val latitude = call.argument<Double>("latitude")
    val longitude = call.argument<Double>("longitude")

    val title = call.argument<String>("title")
    val body = call.argument<String>("body")

    val titleFontSize = call.argument<Double?>("titleFontSize")
    val bodyFontSize = call.argument<Double?>("bodyFontSize")

    val offsetX = call.argument<Double?>("offsetX") ?: 0.0
    val offsetY = call.argument<Double?>("offsetY") ?: 0.0

    if (viewAnnotationId != null && latitude != null && longitude != null && title != null && body != null) {
      // Inflate the custom view layout
      val inflater = LayoutInflater.from(context)
      val customView: View = inflater.inflate(R.layout.view_annotation_layout, mapView, false)

      // Set the custom text on the view
      val titleTextView = customView.findViewById<TextView>(R.id.annotation_title_text)
      titleTextView.text = title

      if (titleFontSize != null) {
        titleTextView.textSize = titleFontSize.toFloat()
      }

      val bodyTextView = customView.findViewById<TextView>(R.id.annotation_body_text)
      bodyTextView.text = body

      if (bodyFontSize != null) {
        bodyTextView.textSize = bodyFontSize.toFloat()
      }

      // Add the view annotation to the map
      viewAnnotationManager.addViewAnnotation(
        customView,
        viewAnnotationOptions {
          geometry(Point.fromLngLat(longitude, latitude))
          annotationAnchor {
            anchor(ViewAnnotationAnchor.BOTTOM)
            offsetX(offsetX)
            offsetY(offsetY)
          }
          allowOverlap(true)
        }
      )

      // Store the annotation view by ID
      annotationsMap[viewAnnotationId] = customView

      result.success(null)
    } else {
      result.error("INVALID_ARGUMENTS", "Latitude, longitude, or text is null", null)
    }
  }

  fun updateViewAnnotation(call: MethodCall, result: MethodChannel.Result) {
    val viewAnnotationId = call.argument<String>("viewAnnotationId")
    val latitude = call.argument<Double>("latitude")
    val longitude = call.argument<Double>("longitude")

    val title = call.argument<String>("title")
    val body = call.argument<String>("body")

    val titleFontSize = call.argument<Double?>("titleFontSize")
    val bodyFontSize = call.argument<Double?>("bodyFontSize")

    if (viewAnnotationId != null && latitude != null && longitude != null && title != null && body != null) {
      val customView = annotationsMap[viewAnnotationId]

      if (customView != null) {
        viewAnnotationManager.updateViewAnnotation(
          customView,
          viewAnnotationOptions {
            geometry(Point.fromLngLat(longitude, latitude))
          }
        )

        val titleTextView = customView.findViewById<TextView>(R.id.annotation_title_text)
        titleTextView.text = title

        if (titleFontSize != null) {
          titleTextView.textSize = titleFontSize.toFloat()
        }

        val bodyTextView = customView.findViewById<TextView>(R.id.annotation_body_text)
        bodyTextView.text = body

        if (bodyFontSize != null) {
          bodyTextView.textSize = bodyFontSize.toFloat()
        }

        result.success(null)
      } else {
        result.error("ANNOTATION_NOT_FOUND", "No annotation found with id $viewAnnotationId", null)
      }
    } else {
      result.error("INVALID_ARGUMENTS", "ID or text is null", null)
    }
  }

  fun removeViewAnnotation(call: MethodCall, result: MethodChannel.Result) {
    val viewAnnotationId = call.argument<String>("viewAnnotationId")

    if (viewAnnotationId != null) {
      val customView = annotationsMap[viewAnnotationId]
      if (customView != null) {
        viewAnnotationManager.removeViewAnnotation(customView)
        annotationsMap.remove(viewAnnotationId)
        result.success(null)
      } else {
        result.error("ANNOTATION_NOT_FOUND", "No annotation found with id $viewAnnotationId", null)
      }
    } else {
      result.error("INVALID_ARGUMENTS", "ID is null", null)
    }
  }

  fun viewAnnotationExists(call: MethodCall, result: MethodChannel.Result) {
    val viewAnnotationId = call.argument<String>("viewAnnotationId")

    if (viewAnnotationId != null) {
      val customView = annotationsMap[viewAnnotationId]

      result.success(customView != null)
    } else {
      result.error("INVALID_ARGUMENTS", "ID is null", null)
    }
  }
}

File: ios/Classes/ViewAnnotationController.swift
import Flutter
import UIKit
import MapboxMaps

class ViewAnnotationController: NSObject {

    private var mapView: MapView
    private var annotationsMap: [String: UIView] = [:] // Store annotations by their ID

    init(withMapView mapView: MapView) {
        self.mapView = mapView
        super.init()
    }

    // Add view annotation to the map
    func addViewAnnotation(methodCall: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = methodCall.arguments as? [String: Any],
              let viewAnnotationId = args["viewAnnotationId"] as? String,
              let latitude = args["latitude"] as? Double,
              let longitude = args["longitude"] as? Double,
              let title = args["title"] as? String,
              let body = args["body"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENTS", message: "Missing arguments", details: nil))
            return
        }

        // Optional offsets with defaults
        let offsetX = args["offsetX"] as? Double ?? 0.0
        let offsetY = args["offsetY"] as? Double ?? 0.0
        
        // Optional font sizes with defaults
        let titleFontSize = args["titleFontSize"] as? Double ?? 16
        let bodyFontSize = args["bodyFontSize"] as? Double ?? 14

        // Check if the annotation already exists
        if annotationsMap[viewAnnotationId] != nil {
            result(FlutterError(code: "ANNOTATION_EXISTS", message: "Annotation with this ID already exists", details: nil))
            return
        }

        // Create a container view without a background
        let annotationView = createAnnotationView(title: title, body: body, titleFontSize: titleFontSize, bodyFontSize: bodyFontSize)

        // Calculate dynamic size
        annotationView.layoutIfNeeded()
        let dynamicSize = annotationView.systemLayoutSizeFitting(UIView.layoutFittingCompressedSize)

        // Add annotation to Mapbox
        let centerCoordinate = CLLocationCoordinate2D(latitude: latitude, longitude: longitude)
        let point = Point(centerCoordinate)

        let viewAnnotationOptions = ViewAnnotationOptions(
            geometry: point,
            width: dynamicSize.width,
            height: dynamicSize.height,
            allowOverlap: true,
            offsetX: offsetX,
            offsetY: offsetY
        )

        do {
            try mapView.viewAnnotations.add(annotationView, options: viewAnnotationOptions)
            annotationsMap[viewAnnotationId] = annotationView
            result(nil)
        } catch {
            result(FlutterError(code: "ANNOTATION_ERROR", message: "Failed to add annotation", details: error.localizedDescription))
        }
    }

    // Update view annotation with new content
    func updateViewAnnotation(methodCall: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = methodCall.arguments as? [String: Any],
              let viewAnnotationId = args["viewAnnotationId"] as? String,
              let title = args["title"] as? String,
              let body = args["body"] as? String,
              let latitude = args["latitude"] as? Double,
              let longitude = args["longitude"] as? Double else {
            result(FlutterError(code: "INVALID_ARGUMENTS", message: "Missing arguments", details: nil))
            return
        }

        guard let annotationView = annotationsMap[viewAnnotationId] else {
            result(FlutterError(code: "ANNOTATION_NOT_FOUND", message: "Annotation not found", details: nil))
            return
        }

        // Optional font sizes with defaults
        let titleFontSize = args["titleFontSize"] as? Double
        let bodyFontSize = args["bodyFontSize"] as? Double
        
        // Update title and body
        if let titleLabel = annotationView.subviews.first(where: { $0.accessibilityIdentifier == "titleLabel" }) as? UILabel {
            titleLabel.text = title
            if let fontSize = titleFontSize { // Safely unwrap the optional
                titleLabel.font = .boldSystemFont(ofSize: CGFloat(fontSize))
            }
        }

        if let bodyLabel = annotationView.subviews.first(where: { $0.accessibilityIdentifier == "bodyLabel" }) as? UILabel {
            bodyLabel.text = body
            if let fontSize = bodyFontSize { // Safely unwrap the optional
                bodyLabel.font = .systemFont(ofSize: CGFloat(fontSize))
            }
        }

        // Calculate dynamic size after updates
        annotationView.layoutIfNeeded()
        let dynamicSize = annotationView.systemLayoutSizeFitting(UIView.layoutFittingCompressedSize)

        // Update the annotation options
        let centerCoordinate = CLLocationCoordinate2D(latitude: latitude, longitude: longitude)
        let point = Point(centerCoordinate)

        let viewAnnotationOptions = ViewAnnotationOptions(
            geometry: point,
            width: dynamicSize.width,
            height: dynamicSize.height,
            allowOverlap: true
        )

        do {
            try mapView.viewAnnotations.update(annotationView, options: viewAnnotationOptions)
            result(nil)
        } catch {
            result(FlutterError(code: "ANNOTATION_ERROR", message: "Failed to update annotation", details: error.localizedDescription))
        }
    }

    // Remove view annotation from the map
    func removeViewAnnotation(methodCall: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = methodCall.arguments as? [String: Any],
              let viewAnnotationId = args["viewAnnotationId"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENTS", message: "Missing arguments", details: nil))
            return
        }

        guard let annotationView = annotationsMap[viewAnnotationId] else {
            result(FlutterError(code: "ANNOTATION_NOT_FOUND", message: "Annotation not found", details: nil))
            return
        }

        do {
            try mapView.viewAnnotations.remove(annotationView)
            annotationsMap.removeValue(forKey: viewAnnotationId)
            result(nil)
        } catch {
            result(FlutterError(code: "ANNOTATION_ERROR", message: "Failed to remove annotation", details: error.localizedDescription))
        }
    }

    // Check if an annotation exists
    func viewAnnotationExists(methodCall: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = methodCall.arguments as? [String: Any],
              let viewAnnotationId = args["viewAnnotationId"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENTS", message: "Missing arguments", details: nil))
            return
        }

        result(annotationsMap[viewAnnotationId] != nil)
    }

    // Helper method to create the annotation view
    private func createAnnotationView(title: String, body: String, titleFontSize: Double, bodyFontSize: Double) -> UIView {
        let annotationView = UIView()
        annotationView.translatesAutoresizingMaskIntoConstraints = false

        // Title label
        let titleLabel = UILabel()
        titleLabel.accessibilityIdentifier = "titleLabel"
        titleLabel.text = title
        titleLabel.font = .boldSystemFont(ofSize: CGFloat(titleFontSize))
        titleLabel.textColor = UIColor.green
        titleLabel.textAlignment = .center
        titleLabel.translatesAutoresizingMaskIntoConstraints = false
        annotationView.addSubview(titleLabel)

        // Body label
        let bodyLabel = UILabel()
        titleLabel.accessibilityIdentifier = "bodyLabel"
        bodyLabel.text = body
        bodyLabel.font = .boldSystemFont(ofSize:  CGFloat(bodyFontSize))
        bodyLabel.textColor = UIColor.green
        bodyLabel.numberOfLines = 0
        bodyLabel.translatesAutoresizingMaskIntoConstraints = false
        annotationView.addSubview(bodyLabel)

        // Constraints
        NSLayoutConstraint.activate([
            titleLabel.topAnchor.constraint(equalTo: annotationView.topAnchor),
            titleLabel.leadingAnchor.constraint(equalTo: annotationView.leadingAnchor),
            titleLabel.trailingAnchor.constraint(equalTo: annotationView.trailingAnchor),

            bodyLabel.topAnchor.constraint(equalTo: titleLabel.bottomAnchor, constant: 4),
            bodyLabel.leadingAnchor.constraint(equalTo: annotationView.leadingAnchor),
            bodyLabel.trailingAnchor.constraint(equalTo: annotationView.trailingAnchor),
            bodyLabel.bottomAnchor.constraint(equalTo: annotationView.bottomAnchor)
        ])

        return annotationView
    }
}
